node/forest_libp2p/src/rpc/mod.rs

```diff
@@ -142,12 +118,45 @@ where
     where
         T: AsyncWrite + Unpin + Send,
     {
-        // TODO: Use FramedWrite to stream write. Dilemma right now is if we should fork the cbor codec so we can replace serde_cbor to our fork of serde_cbor
-        io.write_all(
-            &to_vec(&res).map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?,
-        )
-        .await?;
-        io.close().await?;
-        Ok(())
+        encode_and_write(io, res).await
     }
 }
+
+async fn read_and_decode<IO, T>(io: &mut IO) -> io::Result<T>
+where
+    IO: AsyncRead + Unpin,
+    T: serde::de::DeserializeOwned,
+{
+    const MAX_BYTES_ALLOWED: usize = 2 * 1024 * 1024; // messages over 2MB are likely malicious
+    const TIMEOUT_SECS: u64 = 30;
+
+    // Currently the protocol does not send length encoded message,
+    // and we use `decode-success-with-no-trailing-data` to detect end of frame
+    // just like what `FramedRead` does, so it's possible to cause deadlock at `io.poll_ready`
+    // Adding timeout here to mitigate the issue
+    match tokio::time::timeout(
+        Duration::from_secs(TIMEOUT_SECS),
+        DagCborDecodingReader::new(io, MAX_BYTES_ALLOWED),
+    )
+    .await
+    {
+        Ok(r) => r,
+        Err(_) => {
+            let err = io::Error::new(io::ErrorKind::Other, "read_and_decode timeout");
+            log::warn!("{err}");
+            Err(err)
+        }
+    }
+}
+
+async fn encode_and_write<IO, T>(io: &mut IO, data: T) -> io::Result<()>
+where
+    IO: AsyncWrite + Unpin,
+    T: serde::Serialize,
+{
+    let bytes = fvm_ipld_encoding::to_vec(&data)
+        .map_err(|e| io::Error::new(io::ErrorKind::Other, e.to_string()))?;
+    io.write_all(&bytes).await?;
+    io.close().await?;
```

Comment:
do we need to close manually? If yes, what happens in case of an error on the `write_all`?

Reply:
I don't think it's needed, because the framework calls `io.close` already, but this follows the `libp2p` example: https://github.com/libp2p/rust-libp2p/blob/master/examples/file-sharing.rs#L734